subsys/net/lib/sntp_sock/sntp_sock.c

```diff
@@ -0,0 +1,232 @@
+{
+	if (!pkt) {
+		return;
+	}
+
+	NET_DBG("li               %x", LVM_GET_LI(pkt->lvm));
+	NET_DBG("vn               %x", LVM_GET_VN(pkt->lvm));
+	NET_DBG("mode             %x", LVM_GET_MODE(pkt->lvm));
+	NET_DBG("stratum:         %x", pkt->stratum);
+	NET_DBG("poll:            %x", pkt->poll);
+	NET_DBG("precision:       %x", pkt->precision);
+	NET_DBG("root_delay:      %x", pkt->root_delay);
+	NET_DBG("root_dispersion: %x", pkt->root_dispersion);
+	NET_DBG("ref_id:          %x", pkt->ref_id);
+	NET_DBG("ref_tm_s:        %x", pkt->ref_tm_s);
+	NET_DBG("ref_tm_f:        %x", pkt->ref_tm_f);
+	NET_DBG("orig_tm_s:       %x", pkt->orig_tm_s);
+	NET_DBG("orig_tm_f:       %x", pkt->orig_tm_f);
+	NET_DBG("rx_tm_s:         %x", pkt->rx_tm_s);
+	NET_DBG("rx_tm_f:         %x", pkt->rx_tm_f);
+	NET_DBG("tx_tm_s:         %x", pkt->tx_tm_s);
+	NET_DBG("tx_tm_f:         %x", pkt->tx_tm_f);
+}
+
+static s32_t parse_response(u8_t *data, u16_t len, u32_t orig_ts,
+				u64_t *epoch_time)
+{
+	struct sntp_pkt *pkt = (struct sntp_pkt *)data;
+
+	sntp_pkt_dump(pkt);
+
+	if (ntohl(pkt->orig_tm_s) != orig_ts) {
+		NET_DBG("Mismatch originate timestamp: %d, expect: %d",
+			ntohl(pkt->orig_tm_s), orig_ts);
+		return -EINVAL;
+	}
+
+	if (LVM_GET_LI(pkt->lvm) > SNTP_LI_MAX) {
+		NET_DBG("Unexpected LI: %d", LVM_GET_LI(pkt->lvm));
+		return -EINVAL;
+	}
+
+	if (LVM_GET_MODE(pkt->lvm) != SNTP_MODE_SERVER) {
+		/* For unicast and manycast, server should return 4.
+		 * For broadcast (which is not supported now), server should
+		 * return 5.
+		 */
+		NET_DBG("Unexpected mode: %d", LVM_GET_MODE(pkt->lvm));
+		return -EINVAL;
+	}
+
+	if (pkt->stratum == SNTP_STRATUM_KOD) {
+		NET_DBG("kiss-o'-death stratum");
+		return -EBUSY;
+	}
+
+	if (ntohl(pkt->tx_tm_s) == 0 && ntohl(pkt->tx_tm_f) == 0) {
+		NET_DBG("zero transmit timestamp");
+		return -EINVAL;
+	}
+
+	if (epoch_time) {
+		u32_t ts = ntohl(pkt->tx_tm_s);
+
+		/* Check if most significant bit is set */
+		if (ts & 0x80000000) {
+			/* UTC time is reckoned from 0h 0m 0s UTC
+			 * on 1 January 1900.
+			 */
+			if (ts >= OFFSET_1970_JAN_1) {
+				*epoch_time = ts - OFFSET_1970_JAN_1;
+			} else {
+				return -EINVAL;
+			}
+		} else {
+			/* UTC time is reckoned from 6h 28m 16s UTC
+			 * on 7 February 2036.
+			 */
+			*epoch_time = ts + 0x100000000 - OFFSET_1970_JAN_1;
+		}
+	}
+
+	return 0;
+}
+
+static int sntp_recv_response(struct sntp_ctx *sntp, u32_t timeout)
+{
+	struct sntp_pkt buf = { 0 };
+	u64_t epoch_time = 0;
+	u64_t tmp = 0;
+	int status;
+	int rcvd;
+
+	if (poll(sntp->sock.fds, sntp->sock.nfds, timeout) < 0) {
+		NET_ERR("Error in poll:%d", errno);
+		status = -errno;
+		goto error_exit;
+	}
+
+	rcvd = recv(sntp->sock.fd, (u8_t *)&buf, sizeof(buf), 0);
+	if (rcvd < 0) {
+		status = -errno;
+		goto error_exit;
+	}
+
+	if (rcvd != sizeof(struct sntp_pkt)) {
+		status = -EMSGSIZE;
+		goto error_exit;
+	}
+
+	status = parse_response((u8_t *)&buf, sizeof(buf),
+				sntp->expected_orig_ts, &tmp);
```

Comment:
I don't think you need tmp at all, just use epoch_time.

If status is not 0, user should not care about epoch_time validity.